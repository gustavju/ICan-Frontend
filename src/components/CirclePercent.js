import React from 'react';

const CirclePercent = ({ percent, percentColor }) => {
    let percentNum = parseInt(percent);
    // 90 is from svg r (radius)
    let strokeDashoffset = ((100 - percentNum) / 100) * (Math.PI * (90 * 2));
    let stroke = percentColor;
    return (
        <div className="circle-percent__label" data-pct={percent}>
            <svg class="circle-percent__svg" width="200" height="200" viewport="0 0 100 100" version="1.1" xmlns="http://www.w3.org/2000/svg">
                <circle r="90" cx="100" cy="100" strokeDasharray="565.48" strokeDashoffset="0"></circle>
                <circle style={{strokeDashoffset, stroke}} r="90" cx="100" cy="100" strokeDasharray="565.48" strokeDashoffset="0"></circle>
            </svg>
        </div>
    );
}

export default CirclePercent;
