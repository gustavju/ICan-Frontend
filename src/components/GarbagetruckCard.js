import React from 'react';

const GarbagetruckCard = ({ selectGarbagetruck, garbagetruck }) => (
    <div onClick={(e) => selectGarbagetruck(e, garbagetruck.garbageTruckId)} className="garbagetruck-card">
        <div className="garbagetruck-card__id">{garbagetruck.garbageTruckId}</div>
        <div className="garbagetruck-card__text"><span>Capacity:</span>{garbagetruck.capacity}</div>
        <div className="garbagetruck-card__text"><span>Location:</span>{garbagetruck.location.longitude}, {garbagetruck.location.latitude}</div>
    </div>
);

export default GarbagetruckCard;