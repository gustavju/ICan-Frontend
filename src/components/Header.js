import React from 'react';
import { Link, NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { setTrashcans } from '../actions/trashcan';
import { setGarbagetrucks } from '../actions/garbagetruck';

class Header extends React.Component {
    constructor(props) {
        super(props);
        this.interval = setInterval(() => {
            this.updateTrash();
            this.updateGarbagetrucks();
        }, 3000);
    }
    componentWillUnmount() {
        clearInterval(this.interval);
    }
    updateTrash() {
        fetch('http://localhost:8500/getTrashcans').then(response => {
            response.json().then((data) => {
                console.log(data);
                this.props.setTrashcans(data);
            });
        });
    }
    updateGarbagetrucks() {
        fetch('http://localhost:8500/getGarbagetruck').then(response => {
            response.json().then((data) => {
                console.log(data);
                this.props.setGarbagetrucks(data);
            });
        });
    }
    render() {
        return (
            <header className="header">
                <div className="content-container">
                    <div className="header__content">
                        <Link className="header__title" to="/">
                            <h1>ICan</h1>
                        </Link>
                        <NavLink activeClassName="is-active" exact={true} className="header__link" to="/">
                            <h4>Trashcans</h4>
                        </NavLink>
                        <NavLink activeClassName="is-active" className="header__link" to="/garbagetruck">
                            <h4>Garbagetrucks</h4>
                        </NavLink>

                    </div>
                </div>
            </header>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        trashcans: state.trashcans,
        garbagetrucks: state.garbagetrucks,
    }
};
const mapDispatchToProps = (dispatch) => ({
    setTrashcans: (trashcans) => dispatch(setTrashcans(trashcans)),
    setGarbagetrucks: (garbagetrucks) => dispatch(setGarbagetrucks(garbagetrucks))
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);
