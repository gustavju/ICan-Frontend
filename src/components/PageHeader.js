import React from 'react';

const PageHeader = ({ title, subTitle, data }) => (
    <div className="page-header">
        <div className="content-container">
            <h2 className="page-header__title">{title}</h2>
            <h3 className="page-header__subtitle"><span>{data}</span> {subTitle}</h3>
        </div>
    </div>
);

export default PageHeader;
