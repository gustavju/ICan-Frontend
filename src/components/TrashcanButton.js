import React from 'react';

export default class TrashcanButton extends React.Component {
    constructor(props) {
        super(props);
        this.sendCommand = this.sendCommand.bind(this);
    }
    sendCommand(e, trashcanId) {
        const target = e.target.value;
        const fetchUrl = `http://localhost:8500/trashcanCommand?command=${target}&id=${trashcanId}`;
        fetch(fetchUrl).then(res => {
            res.json().then((data) => console.log(data.Message));
        });
    }
    render() {
        return (
            <button className="trashcan-card__button" value={this.props.command} onClick={(e) => this.sendCommand(e, this.props.trashcanId)}>{this.props.command.substring(0, 1)}</button>
        );
    }
}