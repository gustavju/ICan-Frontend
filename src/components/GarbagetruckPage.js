import React from 'react';
import { connect } from 'react-redux';
import GoogleMapReact from 'google-map-react';
import FontAwesome from 'react-fontawesome';
import TrashcanTable from './TrashcanTable';
import PageHeader from './PageHeader';
import GarbagetruckCard from './GarbagetruckCard';
import { getPercentColor } from '../styles/styleFunctions';

export class GarbagetruckPage extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            selectedTrashcans: [],
            selctedGarbagetruck: {},
            map: '',
            maps: ''
        };
    }
    sendRoute = () => {
        let trashcanQuery = '';
        const selectedTrashcans = this.state.selectedTrashcans;
        for (let i = 0; i < selectedTrashcans.length; i++) {
            trashcanQuery += `t${i}=${selectedTrashcans[i].trashcanId}&`;
        }
        const url = `http://localhost:8500/garbagetruckRoute?garbagetruckId=${this.state.selctedGarbagetruck.garbageTruckId}&num=${selectedTrashcans.length}&${trashcanQuery}`;
        console.log(url);
        fetch(url).then(res => {
            res.json().then((data) => console.log(data.Message));
        });
        this.getGoogleRoute();
    }
    selectTrashcan = (e, id) => {
        const element = e.target;
        if (!element.classList.contains('selected')) {
            element.classList.add("selected");
            let trashcan = this.props.trashcans.find(trashcan => trashcan.trashcanId === id);
            this.setState(() => ({ selectedTrashcans: [...this.state.selectedTrashcans, trashcan] }));
        } else {
            element.classList.remove("selected");
            let filterdTrashcans = this.state.selectedTrashcans.filter(trashcan => !trashcan.trashcanId === id);
            this.setState(() => ({ selectedTrashcans: filterdTrashcans }));
        }
    }
    selectGarbagetruck = (e, id) => {
        const element = e.target;
        if (!element.classList.contains('selected')) {
            element.classList.add("selected");
            let garbagetruck = this.props.garbagetrucks.find(garbagetruck => garbagetruck.garbageTruckId === id);
            this.setState(() => ({ selctedGarbagetruck: garbagetruck }));
        } else {
            element.classList.remove("selected");
            this.setState(() => ({ selctedGarbagetruck: '' }));
        }
    }
    getGoogleRoute = () => {
        let map = this.state.map;
        let maps = this.state.maps;
        const directionsService = new maps.DirectionsService();
        const directionsDisplay = new maps.DirectionsRenderer();
        let waypoints = this.state.selectedTrashcans.map(trashcan => ({ location: new google.maps.LatLng(trashcan.location.latitude, trashcan.location.longitude) }))
        directionsService.route({
            origin: new google.maps.LatLng(this.state.selctedGarbagetruck.location.latitude, this.state.selctedGarbagetruck.location.longitude),
            destination: new google.maps.LatLng(59.383689, 17.938740),
            waypoints: waypoints,
            optimizeWaypoints: true,
            travelMode: 'DRIVING'
        }, (response, status) => {
            if (status === 'OK') {
                console.log(response);
                directionsDisplay.setDirections(response);
                const routePolyline = new google.maps.Polyline({ path: response.routes[0].overview_path });
                routePolyline.setMap(map);
            } else {
                console.log('Directions request failed due to ' + status);
            }
        });
    };
    render() {
        return (
            <div>
                <PageHeader title="Garbagetrucks" subTitle={"Connected"} data={this.props.garbagetrucks.length} />
                <GoogleMapReact
                    style={{ width: '100%', height: '400px', position: 'relative' }}
                    bootstrapURLKeys={{ key: process.env.GOOGLE_MAPS_API_KEY }}
                    defaultCenter={{ lat: 59.407859, lng: 17.944644 }}
                    defaultZoom={13}
                    onGoogleApiLoaded={({ map, maps }) => this.setState({ map, maps })}
                    yesIWantToUseGoogleMapApiInternals
                >
                    {
                        this.props.garbagetrucks.length > 0 ?
                            this.props.garbagetrucks.map(garbagetruck => {
                                let markerStyles = {
                                    fontSize: '2rem',
                                    color: 'grey'
                                };
                                return (
                                    <FontAwesome style={markerStyles}
                                        key={garbagetruck.garbageTruckId}
                                        lat={garbagetruck.location.latitude}
                                        lng={garbagetruck.location.longitude}
                                        text={garbagetruck.garbageTruckId}
                                        name="truck" >
                                    </FontAwesome>
                                );
                            }
                            ) : <p>No connected garbagetrucks!</p>
                    }
                    {
                        this.props.trashcans.length > 0 ?
                            this.props.trashcans.map(trashcan => {
                                let color = trashcan.isConnected == 'false' ? 'grey' : getPercentColor(trashcan.TrashcanHistoryEntry.trashLevel);
                                let markerStyles = {
                                    fontSize: '2rem',
                                    color
                                };
                                return (
                                    <FontAwesome style={markerStyles}
                                        key={trashcan.trashcanId}
                                        lat={trashcan.location.latitude}
                                        lng={trashcan.location.longitude}
                                        text={trashcan.trashcanId}
                                        name="map-marker" >
                                    </FontAwesome>
                                );
                            }
                            ) : <p>No connected trashcans!</p>
                    }
                </GoogleMapReact>
                <div className="content-container">
                    <button className="button" onClick={this.sendRoute}>Send route</button>
                    <div>{this.props.garbagetrucks.length > 0 && this.props.garbagetrucks.map(garbagetruck => <GarbagetruckCard key={garbagetruck.garbageTruckId} selectGarbagetruck={this.selectGarbagetruck} garbagetruck={garbagetruck} />)}</div>
                    <TrashcanTable selectTrashcan={this.selectTrashcan} trashcans={this.props.trashcans} />
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        trashcans: state.trashcans,
        garbagetrucks: state.garbagetrucks,
    }
};

export default connect(mapStateToProps, undefined)(GarbagetruckPage);