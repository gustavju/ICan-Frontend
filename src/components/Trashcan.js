import React from 'react';
import TrashcanButton from './TrashcanButton';

const commands = [
    'empty',
    'addTrash',
    'startTrashFire'
];

const Trashcan = ({trashcan}) => (
    <div className="trashcan-card">
        <div className="trashcan-card__id">ID: {trashcan.trashcanId}</div>
        <div className="trashcan-card__trashlevel">Trashlevel: {trashcan.TrashcanHistoryEntry.trashLevel}%</div>
        <hr/>
        <div className="trashcan-card__temperature">Temperature: {trashcan.TrashcanHistoryEntry.temperature}Â°C</div>
        <div className="trashcan-card__text"><span>Location:</span> {trashcan.location.longitude}, {trashcan.location.latitude}</div>
        <div className="trashcan-card__text"><span>Last Emptied:</span> {trashcan.TrashcanHistoryEntry.lastEmptied}</div>
        <div className="trashcan-card__text"><span>Data updated:</span> {trashcan.TrashcanHistoryEntry.timestamp}</div>
        <div className="trashcan-card__button-container">
            {commands.map((command) => <TrashcanButton key={command} trashcanId={trashcan.trashcanId} command={command} />)}
        </div>
    </div>
);

export default Trashcan;
