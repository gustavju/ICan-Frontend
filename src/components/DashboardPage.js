import React from 'react';
import Trashcan from './Trashcan';
import PageHeader from './PageHeader';
import GoogleMapReact from 'google-map-react';
import FontAwesome from 'react-fontawesome';
import { connect } from 'react-redux';
import { getPercentColor } from '../styles/styleFunctions';

export class DashboardPage extends React.Component {
    constructor(props) {
        super(props);
    }
    handleMouseEnter() {

    }
    handleMouseLeave() {

    }
    render() {
        return (
            <div>
                <PageHeader title="Trashcans" subTitle={"Connected"} data={this.props.trashcans.length} />
                <GoogleMapReact
                    style={{ width: '100%', height: '400px', position: 'relative' }}
                    bootstrapURLKeys={{ key: process.env.GOOGLE_MAPS_API_KEY }}
                    defaultCenter={{ lat: 59.407859, lng: 17.944644 }}
                    defaultZoom={13}
                    onChildMouseEnter={this.handleMouseEnter}
                    onChildMouseLeave={this.handleMouseLeave}
                >
                    {
                        this.props.trashcans.length > 0 ?
                            this.props.trashcans.map(trashcan => {
                                let color = getPercentColor(trashcan.TrashcanHistoryEntry.trashLevel);
                                if (trashcan.isConnected == 'false') color = 'grey';
                                let markerStyles = {
                                    fontSize: '2rem',
                                    color
                                };
                                return (
                                    <FontAwesome style={markerStyles}
                                        key={trashcan.trashcanId}
                                        lat={trashcan.location.latitude}
                                        lng={trashcan.location.longitude}
                                        text={trashcan.trashcanId}
                                        name="map-marker" >
                                    </FontAwesome>
                                );
                            }
                            ) : <p>No connected trashcans!</p>
                    }
                </GoogleMapReact>
                <div className="content-container">
                    <div className="trashcan-card__container">
                        {
                            this.props.trashcans.length > 0 ?
                                this.props.trashcans.map(trashcan => {
                                    let color = getPercentColor(trashcan.TrashcanHistoryEntry.trashLevel);
                                    return (
                                        <Trashcan percentColor={color} key={trashcan.trashcanId} trashcan={trashcan} />
                                    )
                                })
                                :
                                <p>No connected trashcans!</p>
                        }
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        trashcans: state.trashcans
    }
};

export default connect(mapStateToProps, undefined)(DashboardPage);