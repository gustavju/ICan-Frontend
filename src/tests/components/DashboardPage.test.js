import React from 'react';
import { shallow } from 'enzyme';
import { DashboardPage } from '../../components/DashboardPage';
import trashcans from '../fixtures/trashcans';


test('should render the dashboardpage', () => {
    const wrapper = shallow(<DashboardPage trashcans={trashcans} />);
    expect(wrapper).toMatchSnapshot();
});