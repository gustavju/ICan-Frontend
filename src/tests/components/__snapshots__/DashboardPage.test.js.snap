// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the dashboardpage 1`] = `
<div>
  <PageHeader
    data={2}
    subTitle="Connected"
    title="Trashcans"
  />
  <GoogleMap
    bootstrapURLKeys={
      Object {
        "key": undefined,
      }
    }
    debounced={true}
    defaultCenter={
      Object {
        "lat": 59.407859,
        "lng": 17.944644,
      }
    }
    defaultZoom={13}
    distanceToMouse={[Function]}
    googleMapLoader={[Function]}
    heatmap={Object {}}
    heatmapLibrary={false}
    hoverDistance={30}
    layerTypes={Array []}
    onChildMouseEnter={[Function]}
    onChildMouseLeave={[Function]}
    options={[Function]}
    style={
      Object {
        "height": "400px",
        "position": "relative",
        "width": "100%",
      }
    }
    yesIWantToUseGoogleMapApiInternals={false}
  >
    <FontAwesome
      key="8c13a125-cb3f-4a81-8cc5-c64574b0761f"
      lat="59.398546"
      lng="17.947002"
      name="map-marker"
      style={
        Object {
          "color": "#E74C3C",
          "fontSize": "2rem",
        }
      }
      text="8c13a125-cb3f-4a81-8cc5-c64574b0761f"
    />
    <FontAwesome
      key="0b394d27-1d7a-448e-b7f6-fa38d2f1ea41"
      lat="59.407859"
      lng="17.944644"
      name="map-marker"
      style={
        Object {
          "color": "#52BE80",
          "fontSize": "2rem",
        }
      }
      text="0b394d27-1d7a-448e-b7f6-fa38d2f1ea41"
    />
  </GoogleMap>
  <div
    className="content-container"
  >
    <div
      className="trashcan-card__container"
    >
      <Trashcan
        key="8c13a125-cb3f-4a81-8cc5-c64574b0761f"
        percentColor="#E74C3C"
        trashcan={
          Object {
            "TrashcanHistoryEntry": Object {
              "canStatus": "ONFIRE",
              "lastEmptied": "Mon Dec 31 14:18:38 CET 2018",
              "temperature": "128.0",
              "timestamp": "Mon Dec 31 14:45:08 CET 2018",
              "trashLevel": "100.0",
            },
            "isConnected": "true",
            "location": Object {
              "latitude": "59.398546",
              "longitude": "17.947002",
            },
            "trashcanId": "8c13a125-cb3f-4a81-8cc5-c64574b0761f",
            "type": "household",
          }
        }
      />
      <Trashcan
        key="0b394d27-1d7a-448e-b7f6-fa38d2f1ea41"
        percentColor="#52BE80"
        trashcan={
          Object {
            "TrashcanHistoryEntry": Object {
              "canStatus": "EMPTY",
              "lastEmptied": "Mon Dec 31 14:49:51 CET 2018",
              "temperature": "10.0",
              "timestamp": "Mon Dec 31 14:49:52 CET 2018",
              "trashLevel": "10.0",
            },
            "isConnected": "true",
            "location": Object {
              "latitude": "59.407859",
              "longitude": "17.944644",
            },
            "trashcanId": "0b394d27-1d7a-448e-b7f6-fa38d2f1ea41",
            "type": "electronics",
          }
        }
      />
    </div>
  </div>
</div>
`;
